### PR TITLE
Adds a pseudonym to clang"s windows mangler...

### PR DESCRIPTION
**Pull request by memory-thrasher**
…to handle template argument values that are pointers one-past-the-end of a non-array symbol. Also improves error messages in other template argument scenarios where clang bails.

https://github.com/llvm/llvm-project/issues/97756

I don't think I hooked up the unit test right. I'm not sure one is really needed for what boils down to a tweaked if statement. Please advise.

### CODE REVIEW

**Comment by memory-thrasher** [comment]
Oh looks like that new unit test is getting run. The test execution gave me the command to run them. Working on it now.

**Comment by memory-thrasher** [comment]
now the name of the symbol that is being passed is included in the mangle so different value's opte pointers do not collide. Fallback to empty.

**Comment by memory-thrasher** [comment]
@bolshakov-a @eefriedman @zygoloid 
hoping yall are the right reviewers to tag?

**Comment by MaxEW707** [comment]
@memory-thrasher

Godbolt for reference: https://godbolt.org/z/b9v8KhPET

I don't follow that MSVC 1920+ does not support this mangling properly. It appears that it does. I haven't dug too deep into the C++20 MSVC mangling for NTTP yet but it does appear MSVC does support this properly in its mangling scheme.

I am not opposed to doing our own ad-hoc mangling for now to just get this working but in general `-msvc` triple is supposed to adhere to msvc mangling including all its bugs. I'll try to get to this PR sometime this weekend under the assumption we will use the current custom ad-hoc mangling that isn't MSVC compatible in the interim.

Funnily enough we were just discussing this here, https://github.com/llvm/llvm-project/issues/94650#issuecomment-2209626547, and I was about to write up a discourse on such a discussion tomorrow morning.

**Comment by memory-thrasher** [comment]
> Godbolt for reference: https://godbolt.org/z/b9v8KhPET

Huh. Must be my vstools install on this windows laptop is out of date or broken in some other way. I could copy the scheme in that use case into my PR for at least a non-zero chance that it would match the msvc mangler in all other instances triggering that if. I must say that the thought of digging through the msvc binary to get a comprehensive handling does not feel me with excitement.

> likely with a new triple

I love the idea of another triple though. I would of course be using itanium if interfacing with system resources was not impossible. I'm thinking of two possible interpretations of a hybrid triple:
* msvc for all non-templates, itanium for all templates. Also causes any explicit extern template instances to be ignored to prevent the already remote possibility that a specialization or instance could be provided by a library (which would of course be mangled msvc-style).
* an `__attribute__ ` or other intrinsic to specify (in a header file) that an alternate mangler is to be used for the single connected symbol. As much as I hate the idea of compiler-specific code features, sometimes they are the right answer.

I prefer the former.

**Comment by memory-thrasher** [comment]
PR updated to output mangled function name matching the one from godbolt for this specific use case. I suggest someone better at this stuff than me should eventually attempt to implement a more comprehensive mimicry.

**Comment by rnk** [comment]
I think this NTTP mangling code needs some significant updates, refactoring, and an increase in test coverage. I see cd93532dfc455255cb2fa553090d14aaa52b106b landed last May, probably when I was offline for a while.

**Comment by memory-thrasher** [comment]
> I think this NTTP mangling code needs some significant updates, refactoring, and an increase in test coverage. I see [cd93532](https://github.com/llvm/llvm-project/commit/cd93532dfc455255cb2fa553090d14aaa52b106b) landed last May, probably when I was offline for a while.

Oh for sure. There are 5 other cases in template argument values alone that are still broken, I just made their error messages slightly clearer, at least as to which case is happening. I fixed one specific case I was hitting for my own sake, and am trying to share it but I have little interest in trying to fix the others (for free).

File: clang/lib/AST/MicrosoftMangle.cpp
Comment (bolshakov-a) [inline clang/lib/AST/MicrosoftMangle.cpp:1953]: I think that "lvalue base" and "lvalue path" are clang internal terms. Not sure whether they are worth appearing in diagnostics.

File: clang/test/CodeGen/ms_mangler_templatearg_opte.cpp
Comment (bolshakov-a) [inline clang/test/CodeGen/ms_mangler_templatearg_opte.cpp:1]: I'd prefer something simpler like:
```cpp
struct A { int* pi; };
template <A> void TplFn();
int i;
void Fn() {
  TplFn<A{&i + 1}>();
}
```
with checking of the full mangled name.

File: clang/lib/AST/MicrosoftMangle.cpp
Comment (memory-thrasher) [reply clang/lib/AST/MicrosoftMangle.cpp:1953]: I just want something to differentiate the 5 otherwise identical errors to save some future user that step when reporting or fixing one of them. I'm open for suggests on phrasing, anything other than going back to the same message in all 5 cases.

**Comment by memory-thrasher** [comment]
an unrelated test (Modules/prune.m) is now failing on my end. I'm doing a clean build because I suspect there's something stale in a cache somewhere.

**Comment by memory-thrasher** [comment]
ya it's something with my local build env. same test is now failing when built from the commit before mine. Looks like they passed on the build server though so this should be ready. I fixed the things you mentioned @bolshakov-a . Let me know if you have a different phrasing for those diagnositcs in mind or if you want any other changes made.

**Comment by bolshakov-a** [comment]
LGTM, but I don't have the commit access, hence cannot merge your PR. Someone closely related to the community should take a look.

**Comment by memory-thrasher** [comment]
Alright. Pinging the other reviewers.
@rnk @tahonermann @MaxEW707 @zmodem

**Review by rnk — COMMENTED** [review]
Sorry, I made this comment some time ago, but it looks like I never published it

File: clang/lib/AST/MicrosoftMangle.cpp
Comment (memory-thrasher) [reply clang/lib/AST/MicrosoftMangle.cpp]: Done. I couldn't stand to refactor just one piece of the class so I did all the "cannot mange * yet" errors.
@rnk

**Comment by memory-thrasher** [comment]
I finally got the tests to behave.

**Review by rnk — APPROVED** [review]
Thanks, looks good, I'll wait for premerge checks to come back green before I merge it.

**Pull request merged**
